### ISSUE TITLE
Hostname matching doesn't strip trailing '.' when connecting with proxy

### ISSUE THREAD

**Issue by el-deano** [urllib3/urllib3#2244]
### Subject

Same issue as https://github.com/urllib3/urllib3/issues/1254 / https://github.com/urllib3/urllib3/pull/1255 only when using a proxy.  

### Environment

```
>>> print("OS", platform.platform())
OS Linux-3.10.0-1160.15.2.el7.x86_64-x86_64-with-glibc2.17
>>> print("Python", platform.python_version())
Python 3.9.0
>>> print("urllib3", urllib3.__version__)
urllib3 1.26.4
>>>
```

### Steps to Reproduce
Substitute "someproxy-server.com:8080" with a real proxy server to test this.
Just like https://github.com/urllib3/urllib3/issues/1254 the connection works with github.com but fails with github.com. (trailing dot).

```
import urllib3
proxy = urllib3.ProxyManager('http://someproxy-server.com:8080') 
# this works
proxy.request('GET', 'https://github.com')
# this fails
proxy.request('GET', 'https://github.com.')
```

### Expected Behavior

The trailing dot for domain when connecting via proxy should be stripped just as it is in a direct connection as fixed in https://github.com/urllib3/urllib3/pull/1255.

Since [`_match_hostname()` is using `server_hostname`](https://github.com/urllib3/urllib3/blob/1.26.4/src/urllib3/connection.py#L464) which, in the case of proxy connections, comes from [self._tunnel_host instead of self.host](https://github.com/urllib3/urllib3/blob/1.26.4/src/urllib3/connection.py#L357-L373), it seems that a `.rstrip(".")` should either be added when `self._tunnel_host` is referenced [here](https://github.com/urllib3/urllib3/blob/1.26.4/src/urllib3/connection.py#L371) or when it is constructed from self._proxy_host [here](https://github.com/urllib3/urllib3/blob/1.26.4/src/urllib3/connectionpool.py#L959) or when self,_proxy_host is defined and mutated [here](https://github.com/urllib3/urllib3/blob/1.26.4/src/urllib3/connectionpool.py#L79).  The first option seems the safest, but the last option may be most correct.

### Actual Behavior
```
>>> import urllib3                                                                                                                                                                                               >>> proxy = urllib3.ProxyManager('http://someproxy-server.com:8080')
>>> proxy.request('GET', 'https://github.com')
<urllib3.response.HTTPResponse object at 0x7fce2ad964c0>
>>> proxy.request('GET', 'https://github.com.')
Traceback (most recent call last):
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connectionpool.py", line 696, in urlopen
    self._prepare_proxy(conn)
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connectionpool.py", line 964, in _prepare_proxy
    conn.connect()
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connection.py", line 464, in connect
    _match_hostname(cert, self.assert_hostname or server_hostname)
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connection.py", line 512, in _match_hostname
    match_hostname(cert, asserted_hostname)
  File "/usr/local/lib/python3.9/ssl.py", line 416, in match_hostname
    raise CertificateError("hostname %r "
ssl.SSLCertVerificationError: ("hostname 'github.com.' doesn't match either of 'github.com', 'www.github.com'",)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/request.py", line 74, in request
    return self.request_encode_url(
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/request.py", line 96, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/poolmanager.py", line 532, in urlopen
    return super(ProxyManager, self).urlopen(method, url, redirect=redirect, **kw)
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/poolmanager.py", line 375, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connectionpool.py", line 783, in urlopen
    return self.urlopen(
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connectionpool.py", line 783, in urlopen
    return self.urlopen(
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connectionpool.py", line 783, in urlopen
    return self.urlopen(
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/home/mike/bugtest/lib/python3.9/site-packages/urllib3/util/retry.py", line 574, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='github.com.', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLCertVerificationError("hostname 'github.com.' doesn't match either of 'github.com', 'www.github.com'")))
>>>
```

**Comment by achapkowski** [comment]
I have a similar issue with proxies and connecting to any site using urllib3.  These current fixes do not resolve the issues.

Here is my full stack trace:

```python
File \lib\site-packages\requests\sessions.py, in get:
Line 555:   return self.request('GET', url, **kwargs)

File \lib\site-packages\requests\sessions.py, in request:
Line 542:   resp = self.send(prep, **send_kwargs)

File \lib\site-packages\requests\sessions.py, in send:
Line 655:   r = adapter.send(request, **kwargs)

File \lib\site-packages\requests\adapters.py, in send:
Line 449:   timeout=timeout

File \lib\site-packages\urllib3\connectionpool.py, in urlopen:
Line 696:   self._prepare_proxy(conn)

File \lib\site-packages\urllib3\connectionpool.py, in _prepare_proxy:
Line 964:   conn.connect()

File \lib\site-packages\urllib3\connection.py, in connect:
Line 359:   conn = self._connect_tls_proxy(hostname, conn)

File \lib\site-packages\urllib3\connection.py, in _connect_tls_proxy:
Line 506:   ssl_context=ssl_context,

File \lib\site-packages\urllib3\util\ssl_.py, in ssl_wrap_socket:
Line 432:   ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls)

File \lib\site-packages\urllib3\util\ssl_.py, in _ssl_wrap_socket_impl:
Line 474:   return ssl_context.wrap_socket(sock)

File \lib\ssl.py, in wrap_socket:
Line 423:   session=session

File \lib\ssl.py, in _create:
Line 827:   raise ValueError("check_hostname requires server_hostname")
```

I've tried monkey patching the ssl and urllib hostname checking functions:

```
import ssl, urllib3

urllib3.connection._match_hostname = lambda cert, asserted_hostname: None
ssl.match_hostname = lambda cert, hostname : None
```

Just to see if I could get around this issue, but it still bubbles up.  It seems that when a context is given the proxy does not honor that context.  So if I do something like:

```
context = ssl.create_default_context()
context.check_hostname=False
context.verify_mode = ssl.CERT_NONE
context.load_default_certs()
context.hostname_checks_common_name = False
```

And feed it into the `PoolManager` when a proxy is used, that pool manager seems to be ignored.

My general question is how can I disable this check for proxies?  So if a user refers to the proxy via IP, but the certificate is something like www.myproxy.com, how can I get around this?

**Comment by pquentin** [comment]
@achapkowski Can you please open a new issue with more details about your environment? Your problem seems different, even if it's also a TLS issue involving proxies

**Comment by VaniKulkarni** [comment]
Any updates on this issue, we are using python 3.9.7 and with proxy server, we are not able to build \ pip install packages


[09:39:21]Traceback (most recent call last):


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\cli\base_command.py", line 173, in _main


[09:39:21] status = self.run(options, args)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\cli\req_command.py", line 203, in wrapper


[09:39:21] return func(self, options, args)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\commands\wheel.py", line 140, in run


[09:39:21] requirement_set = resolver.resolve(


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\resolution\resolvelib\resolver.py", line 94, in resolve


[09:39:21] result = self._result = resolver.resolve(


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\resolvelib\resolvers.py", line 472, in resolve


[09:39:21] state = resolution.resolve(requirements, max_rounds=max_rounds)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\resolvelib\resolvers.py", line 341, in resolve


[09:39:21] self._add_to_criteria(self.state.criteria, r, parent=None)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\resolvelib\resolvers.py", line 172, in _add_to_criteria


[09:39:21] if not criterion.candidates:


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\resolvelib\structs.py", line 151, in __bool__


[09:39:21] return bool(self._sequence)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\resolution\resolvelib\found_candidates.py", line 140, in __bool__


[09:39:21] return any(self)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\resolution\resolvelib\found_candidates.py", line 128, in <genexpr>


[09:39:21] return (c for c in iterator if id(c) not in self._incompatible_ids)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\resolution\resolvelib\found_candidates.py", line 29, in _iter_built


[09:39:21] for version, func in infos:


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\resolution\resolvelib\factory.py", line 272, in iter_index_candidate_infos


[09:39:21] result = self._finder.find_best_candidate(


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\index\package_finder.py", line 851, in find_best_candidate


[09:39:21] candidates = self.find_all_candidates(project_name)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\index\package_finder.py", line 798, in find_all_candidates


[09:39:21] page_candidates = list(page_candidates_it)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\index\sources.py", line 134, in page_candidates


[09:39:21] yield from self._candidates_from_page(self._link)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\index\package_finder.py", line 758, in process_project_url


[09:39:21] html_page = self._link_collector.fetch_page(project_url)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\index\collector.py", line 490, in fetch_page


[09:39:21] return _get_html_page(location, session=self.session)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\index\collector.py", line 400, in _get_html_page


[09:39:21] resp = _get_html_response(url, session=session)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\index\collector.py", line 115, in _get_html_response


[09:39:21] resp = session.get(


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\requests\sessions.py", line 555, in get


[09:39:21] return self.request('GET', url, **kwargs)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_internal\network\session.py", line 454, in request


[09:39:21] return super().request(method, url, *args, **kwargs)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\requests\sessions.py", line 542, in request


[09:39:21] resp = self.send(prep, **send_kwargs)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\requests\sessions.py", line 655, in send


[09:39:21] r = adapter.send(request, **kwargs)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\cachecontrol\adapter.py", line 53, in send


[09:39:21] resp = super(CacheControlAdapter, self).send(request, **kw)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\requests\adapters.py", line 439, in send


[09:39:21] resp = conn.urlopen(


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\urllib3\connectionpool.py", line 696, in urlopen


[09:39:21] self._prepare_proxy(conn)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\urllib3\connectionpool.py", line 964, in _prepare_proxy


[09:39:21] conn.connect()


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\urllib3\connection.py", line 359, in connect


[09:39:21] conn = self._connect_tls_proxy(hostname, conn)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\urllib3\connection.py", line 500, in _connect_tls_proxy


[09:39:21] return ssl_wrap_socket(


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\urllib3\util\ssl_.py", line 453, in ssl_wrap_socket


[09:39:21] ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls)


[09:39:21] File "C:\Program Files\Python39\lib\site-packages\pip\_vendor\urllib3\util\ssl_.py", line 495, in _ssl_wrap_socket_impl


[09:39:21] return ssl_context.wrap_socket(sock)


[09:39:21] File "C:\Program Files\Python39\lib\ssl.py", line 500, in wrap_socket


[09:39:21] return self.sslsocket_class._create(


[09:39:21] File "C:\Program Files\Python39\lib\ssl.py", line 997, in _create


[09:39:21] raise ValueError("check_hostname requires server_hostname")


[09:39:21]ValueError: check_hostname requires server_hostname


[09:39:22]Process exited with code 2

**Comment by achapkowski** [comment]
I am using fiddler to monitor some traffic locally on windows 10 machine.

So to mimic my setup:

1. Install fiddler 4
2. Setup to capture the HTTPS traffic
3. Use an IP Address for the Proxy: 127.0.0.1:8888 is the default
3. Send a GET/POST to your favorite website
4. Get the hostname mismatch error (or as I like to say despair lol)

**Comment by pquentin** [comment]
@achapkowski @VaniKulkarni You're getting hit by #2400, can you please stop commenting about it here?

**Comment by zhanpon** [comment]
Tried to fix this issue in #2874, but found out it is more complicated than expected: [A tale of a trailing dot](https://daniel.haxx.se/blog/2022/05/12/a-tale-of-a-trailing-dot).

I also reported an issue related to trailing dots #2873.

**Comment by illia-v** [comment]
> Tried to fix this issue in #2874, but found out it is more complicated than expected: [A tale of a trailing dot](https://daniel.haxx.se/blog/2022/05/12/a-tale-of-a-trailing-dot).

This is a nice article! However, the reported error is related to HTTPS, the article [says](https://daniel.haxx.se/blog/2022/05/12/a-tale-of-a-trailing-dot/#:~:text=and%20vice%20versa.-,SNI,-When%20speaking%20to) that the trailing dot has to be stripped from the hostname in such cases.
